### ISSUE TITLE
Drop down menu not populating with namespaces during secret creation

### ISSUE THREAD

**Issue by gaktive** [rancher/dashboard#8544]
Internal reference: SURE-6067
Reported in 2.6.9 and 2.6.11

Issue description:
A user wanted a custom project "view-only" role with the added ability to create, delete, and edit secrets. We created the custom project role with inherited "view only" role and added secrets with all verb permissions. 

Everything looks good until you go to create a secret for a namespace within the project in the UI. It loads a pre-selected namespace in the project but the drop down menu for namespace is blank even though there are 4 namespaces in the project (see attached screenshots). The project role has get, list, and watch verbs for namespaces.

Business impact:
For this user, app teams are assigned a project. They have issues when app teams create and delete deployments etc. as it causes problems as they are supposed to deploy through their dev portal and not through Rancher. App teams manage (create/delete/edit) secrets via Rancher so we created a specific project role and ran into this problem.

Troubleshooting steps:
We tried several different verbiage for namespace resource in the role to get the drop down menu to populate. It seems like "create" is the one that populates the menu. That is an issue in this use case as this user would like namespaces to be "view only".

Repro steps:

step 1 Create a customer project role
step 2 In "inherit" add the "view only" role
step 3 Add secrets to role with all verbs (create, delete, get, list, watch, update)
step 4 save
step 5 select a project and assign a user the custom project role
step 6 login as said user 
step 7 goto secrets > create. 
step 8 click the drop down menu

Workaround:

There are two...

1. fill out the secret in the UI then before saving select "edit as yaml" and put in the correct namespace
2. Add the "create" verb to the namespace resource in the custom project role. (Not sure if this one will work in this specific case)

Actual behavior:
namespace drop down menu not populating with namespaces when creating secret even with explicitly having namespace resource in the role with get/list verbs for customer project role

Expected behavior:
namespace drop down menu should populate with namespaces with get/list verbs for namespaces

Additional notes:
There is no way to type the namespace in the in the field. It just says "no match found" and doesn't take it.

Ideally you could type a namespace in and then UI would say something like, "this namespace doesn't exist want me to create it? <y/n>"

**Comment by cnotv** [comment]
This can be related to latest changes on https://github.com/rancher/dashboard/issues/6094

**Comment by cnotv** [comment]
Other cases may be affected due logic duplication and inconsistent through components, as defined in https://github.com/rancher/dashboard/issues/8469

**Comment by gaktive** [comment]
Goal here for UI: find out if this is a permissions issue whereby the user has access to see the namespace.

If it is a permissions issue, this would go to `rancher/rancher`.

**Comment by bisht-richa** [comment]
I have tested the issue in 2.6 and 2.7. But can not reproduce it.
Please check the video below comment tested by me and @cnotv

**Comment by JeffParkes** [comment]
@bisht-richa @cnotv We upgraded to 2.7.2 and problem is still there. Please see internal bug (SURE-6067) for screen capture video I posted.

**Comment by gaktive** [comment]
Reopening if we have to re-reproduce this behaviour. If need be, we can work with @JeffParkes with where he's seeing this.

**Comment by JeffParkes** [comment]
Yes @gaktive let me know and we can setup a screenshare or I can get you whatever logs/config you need. Thanks!

**Comment by cnotv** [comment]
I was checking with @bisht-richa about reproducing the issue and we were able to do so.

It looks like the only filter which is preventing to display the namespace is the logic related to "read only". However this is something which has been required and it seems correct from our end.
What is the logic from your point of view? Is it wrong the information from the server?

```ts
/**
 * Get only namespaces for user with roles "Cluster Member" and "View All Projects"
 * @returns Record<string, true>
 */
const getReadOnlyActiveNamespaces = (namespaces, activeNamespaces) => {
  const readonlyNamespaces = Object
    .values(namespaces)
    .filter(ns => !!ns.links.update). // <-- Mentioned filter
    .map(({ id }) => id);

  return Object.keys(activeNamespaces)
    .filter(ns => readonlyNamespaces.includes(ns))
    .reduce((acc, ns) => ({
      ...acc,
      [ns]: true
    }), {});
};
```

https://github.com/rancher/dashboard/blob/master/shell/store/index.js#L127-L143

**Comment by gaktive** [comment]
@cnotv who is that question directed to? I'm trying to understand where this is blocked on.

**Comment by cnotv** [comment]
To anyone from the 3 BE teams we have. Not so sure which one, cannot frame it down to a specific one tbh.

**Comment by gaktive** [comment]
@cnotv can you file a ticket then in `rancher/rancher` and link back here then since I'm not sure fully what the ask is? I can then find the right team to look at it but I think this goes into Q3.

@nwmac let me know if something else needs to be done here based on your own discussion.

**Comment by cnotv** [comment]
@gaktive do we really need to issue a ticket to talk to someone which develops the backend? Is there not a single channel or person or responsible which we can contact and ask "Is the payload correct in this case?" 
I've to make the question to him not to you 😅

**Comment by cnotv** [comment]
@gaktive anyhow this issue is based on code existing from Rancher 2.6.9 and I'm trying to help the assigned developer to find out a solution.

Could you guys support @bisht-richa as well by talking to someone on your own?

**Comment by cnotv** [comment]
As @nwmac yesterday asked in the talk, which endpoint we are trying to identify, it looks like to be `/v1/namespaces`.
Since we don't have a Swagger, this is the data:

```js
{
  "type": "collection",
  "links": {
    "self": "https://rancher.cnotv.xyz/v1/namespaces"
  },
  "createTypes": {
    "namespace": "https://rancher.cnotv.xyz/v1/namespaces"
  },
  "actions": {},
  "resourceType": "namespace",
  "revision": "98848281",
  "count": 37,
  "data": [
    {
      "id": "c-m-26482m2b",
      "type": "namespace",
      "links": {
        "remove": "https://rancher.cnotv.xyz/v1/namespaces/c-m-26482m2b",
        "self": "https://rancher.cnotv.xyz/v1/namespaces/c-m-26482m2b",
        "update": "https://rancher.cnotv.xyz/v1/namespaces/c-m-26482m2b",  // <-- if this present, the user has restrictions to view only
        "view": "https://rancher.cnotv.xyz/api/v1/namespaces/c-m-26482m2b"
      },
      // [ ... ] Rest of the Namespace payload
    },
    // [ ... ]
    // Filtering the remaining list of Namespaces
  ]
}
```

**Comment by nwmac** [comment]
I don't think this is a backend issue.

The user only has readonly permission on the namespace, because they are not allowed to edit it - but they DO see it, which means they have access to it.

The logic in the `getReadOnlyActiveNamespaces` seems wrong - it is actually getting the namespaces that can be updated - i.e. the ones with an update link. So, these are the ones that the user can write to.

**Comment by cnotv** [comment]
The error is since 2.6.9, so it's not a new implementation.

**Comment by gaktive** [comment]
This isn't a backend issue. Moving to Backlog so we can discuss as a team about what to do here during a planning/grooming session to understand the current state.

**Comment by mantis-toboggan-md** [comment]
I've looked into this a little more and I think we'd need backend changes to correctly filter out namespaces a user can't create resources in.

As an admin:
1. Create a standard user, `su1`
2. Create a project role, `create-secrets`, with permission to create secrets
3. Create a project role, `get-secrets`, with permission to get/list secrets
4. Create a project, `p1`, and add `su1` with the `create-secrets` role
5. Create a couple namespaces, `p1-ns1`, `p1-ns2` in the project
6. Create a project, `p2`, and add `su1` with the `get-secrets` role
7. Create a couple namespaces, `p2-ns1`, `p2-ns2` in the project

In this scenario, ideally, when creating a secret `su1` would see `p1-ns1` and `p1-ns2` in the namespace dropdown. `p2-ns1` and `p2-ns2` would still show up in the header namespace filter. What actually happens is that none of those namespaces are in the `NameNsDescription` and all of them are in the `NamespaceFilter`. 

We don't appear to have a way to determine which of the above namespaces the user has permission to create resources in; they all have the same `links` present and I don't see any other differentiating field we could use.

A partial fix on our end would be to stop attempting to filter out read-only namespaces at all. This would mean the `NameNsDescription` dropdown includes namespaces the user potentially can't actually create the resource in, but it solves the bug described above. If we want to filter out read-only namespaces there we would need some kinda backend change to determine ns permissions with more granularity.

**Comment by mantis-toboggan-md** [comment]
This ties into backend changes requested here: https://github.com/rancher/rancher/issues/41929

**Comment by gaktive** [comment]
Related backend pushed to Q4 so doing the same here and updating SURE.

**Comment by JeffParkes** [comment]
> A partial fix on our end would be to stop attempting to filter out read-only namespaces at all. This would mean the NameNsDescription dropdown includes namespaces the user potentially can't actually create the resource in, but it solves the bug described above. If we want to filter out read-only namespaces there we would need some kinda backend change to determine ns permissions with more granularity.

@mantis-toboggan-md in this use case, the namespace is essentially read only _except_ for secrets. We want the ability for the end user to create/delete and manage secrets. Everything else is read only. Would that be considered a read only namespace? Just trying to understand..

**Comment by gaktive** [comment]
More relevant backend ticket now https://github.com/rancher/rancher/issues/41301 as 41929 was dropped in favour of that. Need to check the state of the more recent ticket though it's in the 2024 Q1 milestone at present.

**Comment by gaktive** [comment]
@nwmac to review here to see what we're blocked on here again.

**Comment by richard-cox** [comment]
linked to https://github.com/rancher/rancher/issues/34731, now unblocked
